sql/src/main/java/io/crate/expression/scalar/SubscriptObjectFunction.java

```diff
@@ -31,32 +31,75 @@
 
+import java.util.ArrayList;
 import java.util.List;
 import java.util.Locale;
 import java.util.Map;
+import java.util.function.Function;
 
+/**
+ * Scalar function to resolve elements inside a map.
+ * Because the function registration does not take the function return type into account, it is not possible
+ * to register the same function with different return types under the same name.
+ * As a workaround, the return type is encoded into the function name by using {@link #getNameForReturnType(DataType)}.
+ */
 public class SubscriptObjectFunction extends Scalar<Object, Map> {
 
-    public static final String NAME = "subscript_obj";
+    private static final String NAME = "subscript_obj";
+    private static final FuncParams FUNCTION_PARAMS = FuncParams
+        .builder(Param.of(ObjectType.untyped()), Param.of(StringType.INSTANCE))
+        .withVarArgs(Param.of(StringType.INSTANCE))
+        .build();
+
     private FunctionInfo info;
 
     public static void register(ScalarFunctionModule module) {
-        module.register(NAME,
-            new BaseFunctionResolver(
-                FuncParams.builder(Param.of(ObjectType.untyped()), Param.of(StringType.INSTANCE)).build()
-            ) {
-                @Override
-                public FunctionImplementation getForTypes(List<DataType> dataTypes) throws IllegalArgumentException {
-                    return new SubscriptObjectFunction(
-                        new FunctionInfo(new FunctionIdent(NAME, dataTypes.subList(0, 2)), DataTypes.UNDEFINED));
-                }
+        for (DataType returnType : DataTypes.allRegisteredTypes()) {
+            module.register(getNameForReturnType(returnType),
+                new BaseFunctionResolver(FUNCTION_PARAMS) {
+                    @Override
+                    public FunctionImplementation getForTypes(List<DataType> dataTypes) throws IllegalArgumentException {
+                        return ofReturnType(returnType, dataTypes);
+                    }
+                });
+        }
+    }
+
+    public static SubscriptObjectFunction ofReturnType(DataType returnType, List<DataType> dataTypes) {
+        return new SubscriptObjectFunction(
+            new FunctionInfo(
+                new FunctionIdent(getNameForReturnType(returnType), dataTypes),
+                returnType));
+    }
+
+    public static String getNameForReturnType(DataType dataType) {
+        String name = NAME;
+        if (dataType.id() != UndefinedType.ID) {
+            String returnTypeName;
+            if (DataTypes.isCollectionType(dataType)) {
+                returnTypeName = ((CollectionType) dataType).getCollectionName();
+            } else {
+                returnTypeName = dataType.getName();
             }
-        );
+            name += "_" + returnTypeName;
+        }
+        return name;
+    }
+
+    public static <T> List<T> buildListOfRootAndPath(T rootElement, List<String> path, Function<String, T> convert) {
```

Comment:
Maybe call this `mapTail(O head, List<I> tail, Function<I, O> mapper)`  and put it into `Lists2` ?

Currently it is somehow generic and yet at the same time very specific. (`rootElement`, `path` imply a concrete use-case). I also wasn't able to deduct from the method name alone what exactly it does.